lib/SIL/SILDeclRef.cpp

```diff
@@ -112,129 +112,100 @@ bool swift::requiresForeignEntryPoint(ValueDecl *vd) {
-  } else if (auto *ed = dyn_cast<EnumElementDecl>(vd)) {
-    assert(kind == Kind::EnumElement
-           && "can only create EnumElement SILDeclRef for enum element");
-    naturalUncurryLevel = ed->hasArgumentType() ? 1 : 0;
-  } else if (isa<DestructorDecl>(vd)) {
-    assert((kind == Kind::Destroyer || kind == Kind::Deallocator)
-           && "can only create destroyer/deallocator SILDeclRef for dtor");
-    naturalUncurryLevel = 0;
-  } else if (isa<ClassDecl>(vd)) {
-    assert((kind == Kind::IVarInitializer || kind == Kind::IVarDestroyer) &&
-           "can only create ivar initializer/destroyer SILDeclRef for class");
-    naturalUncurryLevel = 1;
-  } else if (auto *var = dyn_cast<VarDecl>(vd)) {
-    assert((kind == Kind::GlobalAccessor ||
-            kind == Kind::GlobalGetter ||
-            kind == Kind::StoredPropertyInitializer) &&
-           "can only create GlobalAccessor, GlobalGetter or "
-           "StoredPropertyInitializer SILDeclRef for var");
-
-    naturalUncurryLevel = 0;
+  auto setOrCheck = [&](Kind k, Kind k2, Kind k3) {
+    if (setKind) {
+      kind = k;
+    } else {
+      assert((kind == k || kind == k2 || kind == k3) &&
+             "invalid kind of ValueDecl for this SILDeclRef::Kind");
+      (void)k2;
+      (void)k3;
+    }
+  };
+
+  auto assertNoSet = [=]() {
+    assert(!setKind && "can't set SILDeclRef::Kind for this kind of ValueDecl");
+  };
+
+  switch (vd->getKind()) {
+  case DeclKind::Func:
+    setOrCheck(Kind::Func, none, none);
+    return AnyFunctionRef(cast<FuncDecl>(vd)).getNaturalUncurryLevel();
+  case DeclKind::Constructor:
+    setOrCheck(Kind::Allocator, Kind::Initializer, none);
+    return 1;
+  case DeclKind::EnumElement:
+    setOrCheck(Kind::EnumElement, none, none);
+    return cast<EnumElementDecl>(vd)->hasArgumentType() ? 1 : 0;
+  case DeclKind::Destructor:
+    setOrCheck(Kind::Deallocator, Kind::Destroyer, none);
+    return 0;
+  case DeclKind::Class:
+    assertNoSet();
+    setOrCheck(Kind::IVarInitializer, Kind::IVarDestroyer, none);
+    return 1;
+  case DeclKind::Var: {
+    assertNoSet();
+    setOrCheck(Kind::GlobalAccessor, Kind::GlobalGetter,
+               Kind::StoredPropertyInitializer);
+    auto *var = cast<VarDecl>(vd);
     assert(!var->getDeclContext()->isLocalContext() &&
            "can't reference local var as global var");
     assert(var->hasStorage() && "can't reference computed var as global var");
     (void)var;
-  } else {
-    llvm_unreachable("Unhandled ValueDecl for SILDeclRef");
+    return 0;
   }
-  
-  assert((atUncurryLevel == ConstructAtNaturalUncurryLevel
-          || atUncurryLevel <= naturalUncurryLevel)
-         && "can't emit SILDeclRef below natural uncurry level");
-  uncurryLevel = atUncurryLevel == ConstructAtNaturalUncurryLevel
-    ? naturalUncurryLevel
-    : atUncurryLevel;
-  isCurried = uncurryLevel != naturalUncurryLevel;
+  default:
+    llvm_unreachable("unhandled kind of ValueDecl for SILDeclRef");
```

Comment:
Ok, I think I understand now, what you're saying.
But for a covered switch I have to list all other `DeclKind`s, right? Which means Either `break` from the switch and then `llvm_unreachable` or `llvm_unreachable` instead of `break`.